### DISCUSSION TITLE
Crosscompile for Openwrt 19.07 ramips/mt7620 processor

### DISCUSSION THREAD

**Discussion by Drus566** [c-jimenez/open-ocpp#75] [Q&A] [answered]
**Hello, cool library!** 
I'm new to c++ and don't know СMake very well, so I'm asking for help. Tell me or direct me how I can compile this library with all dependencies on open wrt 19.07 on a ramips/7620 processor. I managed to compile and run "Hello World" on this controller. 
Thanks in advance!

**Comment by c-jimenez** [comment]
Hi,
To be able to compile Open-OCPP first you'll need to have a C++17 compliant cross compiler with the pre-requisites libraries and tools listed here https://github.com/c-jimenez/open-ocpp#pre-requisites.
Usually the cross compilers comes with some bash script to setup the environment :
/my_sdk_path/environment-setup-armv7vet2hf-vfpv4d16-linux-gnueabi
/my_sdk_path/sysroots/x86_64-linux/environment-setup.d/cmake.sh

Once you have called them using the "source" command, you just have to call the : make command.

**Reply by Drus566** [reply]
Thanks for the answer!
Do the pre-requisites libraries need to be compiled from source for the target platform?

**Reply by c-jimenez** [reply]
Yes they (OpenSSL ans SQLite) must compiled for the target platform. 
Since these are very common libraries, they are usually part of your Linux distribution buildsystem, so the build is quite easy.
Maybe your current  Linux image already contains them.

**Reply by Drus566** [reply]
Does the existing CMake file need to be rewritten under the library data? If you just run the makefile file from the SDK of the target platform, after setting up the project's CMake call, will the build fail? It turns out it is necessary to rewrite СMake under a makefile?

**Reply by Drus566** [reply]
And what about third party libraries in 3rdParty directory. Will they also need to be recompiled for the target platform?

**Reply by c-jimenez** [reply]
Normaly you should not have to rewrite anything in the CMake files, even under the 3rd party folder => all the source tree should compile "out of the box" once you have configured your cross compiler (see previous reply).
I tried it with a Yocto generated cross arm compiler with success and someone did succeed to build it for raspberry pie without modifying the CMake files.

**Reply by Drus566** [reply]
Greetings!) 
In general, I prepared the toolchain, the simak command worked, but with the make, there were problems. Can you tell me where to look? Perhaps there is a problem with the sqlite3? I compiled sqlite3 and openssl from source and added to the toolchain.
```
[  2%] Built target helpers
[  2%] Built target GENHDR
[ 25%] Built target websockets
[ 26%] Built target ws
[ 27%] Built target database
[ 27%] Building CXX object src/tools/log/CMakeFiles/log.dir/LogDatabase.cpp.o
/home/d1/IRZ/Crosscompile/OCPPClient/src/tools/log/LogDatabase.cpp: In member function 'void ocpp::log::LogDatabase::log(time_t, unsigned int, const string&, const string&)':
/home/d1/IRZ/Crosscompile/OCPPClient/src/tools/log/LogDatabase.cpp:46:42: error: call of overloaded 'bind(int, time_t&)' is ambiguous
         m_insert_query->bind(0, timestamp);
                                          ^
In file included from /home/d1/IRZ/Crosscompile/OCPPClient/src/tools/log/LogDatabase.h:22:0,
                 from /home/d1/IRZ/Crosscompile/OCPPClient/src/tools/log/LogDatabase.cpp:19:
/home/d1/IRZ/Crosscompile/OCPPClient/src/tools/database/./Database.h:107:14: note: candidate: bool ocpp::database::Database::Query::bind(int, bool)
         bool bind(int number, bool value);
              ^~~~
/home/d1/IRZ/Crosscompile/OCPPClient/src/tools/database/./Database.h:115:14: note: candidate: bool ocpp::database::Database::Query::bind(int, double)
         bool bind(int number, double value);
              ^~~~
/home/d1/IRZ/Crosscompile/OCPPClient/src/tools/database/./Database.h:123:14: note: candidate: bool ocpp::database::Database::Query::bind(int, int32_t)
         bool bind(int number, int32_t value);
              ^~~~
/home/d1/IRZ/Crosscompile/OCPPClient/src/tools/database/./Database.h:131:14: note: candidate: bool ocpp::database::Database::Query::bind(int, uint32_t)
         bool bind(int number, uint32_t value);
              ^~~~
/home/d1/IRZ/Crosscompile/OCPPClient/src/tools/database/./Database.h:139:14: note: candidate: bool ocpp::database::Database::Query::bind(int, int64_t)
         bool bind(int number, int64_t value);
              ^~~~
/home/d1/IRZ/Crosscompile/OCPPClient/src/tools/database/./Database.h:147:14: note: candidate: bool ocpp::database::Database::Query::bind(int, uint64_t)
         bool bind(int number, uint64_t value);
              ^~~~
make[2]: *** [src/tools/log/CMakeFiles/log.dir/build.make:63: src/tools/log/CMakeFiles/log.dir/LogDatabase.cpp.o] Ошибка 1
make[1]: *** [CMakeFiles/Makefile2:3328: src/tools/log/CMakeFiles/log.dir/all] Ошибка 2
make: *** [Makefile:163: all] Ошибка 2
```

**Reply by c-jimenez** [reply]
Hi,
You must have a 32bits compiler. This issue has already been fixed for the raspberry build but it is still in the "develop" branch ([PR#70](https://github.com/c-jimenez/open-ocpp/pull/70)).
Please use this branch for now. I will try to take some time to merge it and to create a release by the end of the week.

**Reply by Drus566** [reply]
I fixed this error by explicitly casting to `(int64_t)` and there was also a misunderstanding of `<filesystem> `, I fixed it by replacing `<filesystem>` with `<experimental/filesystem>` and also `std::filesystem` with `std::experimental::filesystem` but then an error popped up...
```
[ 53%] Built target chargepoint
Scanning dependencies of target open-ocpp-dynamic
[ 53%] Building CXX object CMakeFiles/open-ocpp-dynamic.dir/src/version.cpp.o
[ 54%] Linking CXX shared library ../bin/native/libopen-ocpp.so
/home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/7.5.0/../../../../mipsel-openwrt-linux-musl/bin/ld: /home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/usr/lib/libsqlite3.so:(.sdata+0x0): multiple definition of `__dso_handle'; /home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/7.5.0/crtbeginS.o:(.sdata+0x0): first defined here
/home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/7.5.0/../../../../mipsel-openwrt-linux-musl/bin/ld: /home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/usr/lib/libsqlite3.so: in function `_init':
/builder/shared-workdir/build/build_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/musl-1.1.24/crt/mips/crti.s:8: multiple definition of `_init'; /home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/7.5.0/../../../../mipsel-openwrt-linux-musl/lib/crti.o:/builder/shared-workdir/build/build_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/musl-1.1.24/crt/mips/crti.s:8: first defined here
/home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/7.5.0/../../../../mipsel-openwrt-linux-musl/bin/ld: /home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/usr/lib/libsqlite3.so: in function `_fini':
/builder/shared-workdir/build/build_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/musl-1.1.24/crt/mips/crti.s:17: multiple definition of `_fini'; /home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/7.5.0/../../../../mipsel-openwrt-linux-musl/lib/crti.o:/builder/shared-workdir/build/build_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/musl-1.1.24/crt/mips/crti.s:17: first defined here
/home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/7.5.0/../../../../mipsel-openwrt-linux-musl/bin/ld: /home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/usr/lib/libsqlite3.so:(.got+0x0): multiple definition of `_GLOBAL_OFFSET_TABLE_'; /home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/7.5.0/../../../../mipsel-openwrt-linux-musl/lib/crti.o:(.got+0x0): first defined here
/home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/7.5.0/../../../../mipsel-openwrt-linux-musl/bin/ld: /home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/7.5.0/crtendS.o:(.dtors+0x0): multiple definition of `__DTOR_END__'; /home/d1/IRZ/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-7.5.0_musl/usr/lib/libsqlite3.so:(.dtors+0x4): first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/open-ocpp-dynamic.dir/build.make:279: ../bin/native/libopen-ocpp.so] Ошибка 1
make[1]: *** [CMakeFiles/Makefile2:1344: CMakeFiles/open-ocpp-dynamic.dir/all] Ошибка 2
make: *** [Makefile:163: all] Ошибка 2
```

**Comment by Drus566** [comment, marked as answer]
In general, I made this makefile, and fixed a couple of errors that are googled during the compilation process, and got my build)
```bash
include $(TOPDIR)/rules.mk
 
PKG_NAME:=ocppclient
PKG_VERSION:=06.10.2022
PKG_RELEASE=1
 
PKG_LICENSE:=GPL-2.0
PKG_LICENSE_FILES:=
 
PKG_MAINTAINER:=drus <zxandruxazx56@yandex.ru>
 
include $(INCLUDE_DIR)/package.mk
include $(INCLUDE_DIR)/cmake.mk #If usingcmake, You must add this line
 
define Package/ocppclient
	SECTION:=net
	CATEGORY:=ocppclient
	DEPENDS:=+kmod-tun +libpthread +librt +libubox +libuci +libstdcpp +libsqlite3 #Add the required dependency library, if you don’t add it, it will compile and report an error
	TITLE:=OCPPClient
endef

define Package/ocppclient/extra_provides
	echo "/home/d1/IRZ/openwrt-sdk/staging_dir/target-mipsel_24kc_musl/usr/lib/libcrypto.so.1.1"; \
	echo "/home/d1/IRZ/openwrt-sdk/staging_dir/target-mipsel_24kc_musl/usr/lib/libssl.so.1.1";
endef
 
define Build/Prepare
	mkdir -p $(PKG_BUILD_DIR)
	$(CP) ./SOURCES/* $(PKG_BUILD_DIR)/
endef
 
define Package/occpclient/description
	ocppclient from cmake
endef
 
define Package/ocppclient/install
	$(INSTALL_DIR) $(1)/usr/share/ocppclient
	$(INSTALL_DIR) $(1)/usr/lib
	$(CP) $(PKG_BUILD_DIR)/bin/native/libopen-ocpp.so $(1)/usr/lib
	$(CP) $(PKG_BUILD_DIR)/bin $(1)/usr/share/ocppclient
	$(CP) $(PKG_BUILD_DIR)/schemas $(1)/usr/share/ocppclient
endef
$(eval $(call BuildPackage,ocppclient))
```
Thank you! You can close the topic)

**Reply by c-jimenez** [reply]
Thank you for you feedback, I hope it will helpsome other people working with an OpenWRT distribution.

**Reply by Drus566** [reply]
My bad sorry, THIS WORK)
```
define Package/ocppclient
	SECTION:=net
	CATEGORY:=ocppclient
	DEPENDS:=+kmod-tun +libpthread +librt +libubox +libuci +libstdcpp +libsqlite3 +libopenssl #Add the required dependency library, if you don’t add it, it will compile and report an error
	TITLE:=OCPPClient
endef
```

**Reply by Drus566** [reply]
> Thank you for you feedback, I hope it will helpsome other people working with an OpenWRT distribution.

Greetings, do you have an implementation so that both connectors can be processed at the same time? In /examples only sequential processing
Thanks in advance!

**Reply by c-jimenez** [reply]
Hi,
You can process multiple connectors in // using multiple thread since the stack should be thread safe.
You can also manage all the connectors in // in a single thread like in this example => https://github.com/c-jimenez/open-ocpp-simu/blob/develop/src/chargepoint/SimulatedChargePoint.cpp.

**Comment by wlbleaboy** [comment]
Just set the "cross_compiler path" in  two CMakeLists.txt, which is located in the [top Directory] and in the [src Directory] .
like thus:
in the top Directory's CMakeLists.txt
       cmake_minimum_required(VERSION 3.13)
       SET(TOOLCHAIN_DIR   /opt/cross_compiler)
       SET(CMAKE_C_COMPILER        ${TOOLCHAIN_DIR}/bin/arm-linux-gcc)
       SET(CMAKE_CXX_COMPILER      ${TOOLCHAIN_DIR}/bin/arm-linux-g++)

in the src Directory's CMakeLists.txt
       SET(TOOLCHAIN_DIR   /opt/cross_compiler)
       SET(CMAKE_C_COMPILER        ${TOOLCHAIN_DIR}/bin/arm-linux-gcc)
       SET(CMAKE_CXX_COMPILER      ${TOOLCHAIN_DIR}/bin/arm-linux-g++)

      # Subdirectories
       add_subdirectory(centralsystem)
